File: driver/fake/fake.go
// Package fake provides an in-memory Driver/Session implementation that
// spawns no process. It exists to prove the driver abstraction end-to-end and to
// serve as a test double for hosts: a session can be scripted to emit an
// arbitrary sequence of events per prompt, and resume is modelled by
// pre-assigning a session id.
package fake

import (
	"context"
	"fmt"
	"path/filepath"
	"sync"
	"sync/atomic"

	"github.com/Fullstop000/unio/driver"
)

// Script describes the events a single Send or Respond should emit, in order, before the
// terminal Completed/Failed. If FailWith is non-nil the run ends with a Failed
// event carrying it; otherwise it ends Completed with Result.
type Script struct {
	Items    []driver.AgentEventItem
	Result   driver.RunResult
	FailWith *driver.AgentError
	Blocked  *driver.BlockedReason
	Wait     <-chan struct{}
	// InterruptWait delays interrupt confirmation for state-ordering tests.
	InterruptWait <-chan struct{}
}

// Driver is an in-memory Driver. It mints monotonic session ids and
// hands each opened session an optional queue of Scripts consumed per submission.
type Driver struct {
	ctx            context.Context
	spec           driver.AgentSpec
	mu             sync.Mutex
	seq            atomic.Uint64
	stored         []driver.StoredSessionMeta
	scripts        [][]Script
	probe          driver.ProbeAuth
	probeErr       error
	requireInstall bool
}

// New constructs a fake driver reporting an installed+authed probe by default.
func New(ctx context.Context, spec driver.AgentSpec) *Driver {
	return &Driver{
		ctx: ctx, spec: spec, probe: driver.AuthAuthed,
	}
}

// SetProbe overrides what Probe returns (test knob).
func (d *Driver) SetProbe(p driver.ProbeAuth, err error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.probe = p
	d.probeErr = err
}

// SetRequireInstall makes OpenSession run driver.ResolveExecutable on the spec
// (like a real driver), so tests can exercise the not_installed path. Off by
// default because the fake normally spawns no process.
func (d *Driver) SetRequireInstall(on bool) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.requireInstall = on
}

// SetStoredSessions sets what ListSessions returns (test knob).
func (d *Driver) SetStoredSessions(metas []driver.StoredSessionMeta) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.stored = metas
}

// ScriptNextSession sets the Scripts that the next opened session will replay.
func (d *Driver) ScriptNextSession(scripts ...Script) {
	d.mu.Lock()
	defer d.mu.Unlock()
	d.scripts = append(d.scripts, append([]Script(nil), scripts...))
}

// Probe implements driver.Driver.
func (d *Driver) Probe() (driver.ProbeAuth, error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	return d.probe, d.probeErr
}

// ListSessions implements driver.Driver.
func (d *Driver) ListSessions(params driver.ListSessionsParams) ([]driver.StoredSessionMeta, error) {
	d.mu.Lock()
	defer d.mu.Unlock()
	out := make([]driver.StoredSessionMeta, 0, len(d.stored))
	for _, meta := range d.stored {
		if params.Cwd == "" || filepath.Clean(meta.Cwd) == filepath.Clean(params.Cwd) {
			out = append(out, meta)
		}
	}
	return out, nil
}

// OpenSession implements driver.Driver.
func (d *Driver) OpenSession(params driver.OpenParams) (*driver.SessionAttachment, error) {
	d.mu.Lock()
	require := d.requireInstall
	var scripts []Script
	if len(d.scripts) > 0 {
		scripts = d.scripts[0]
		d.scripts = d.scripts[1:]
	}
	d.mu.Unlock()

	// Real drivers always do this at OpenSession; the fake does it only when
	// asked, so the not_installed contract can be exercised without a process.
	if require {
		if _, err := driver.ResolveExecutable(d.spec); err != nil {
			return nil, err
		}
	}

	bus := driver.NewEventBus()
	s := &session{
		ctx: d.ctx, driver: d, bus: bus, resume: params.ResumeSessionID, scripts: scripts,
	}
	s.state.Store(&driver.ProcessState{Phase: driver.PhaseIdle})
	return &driver.SessionAttachment{Session: s, Events: bus}, nil
}

func (d *Driver) nextSessionID() driver.SessionID {
	return fmt.Sprintf("fake-session-%d", d.seq.Add(1))
}

// session implements driver.Session in memory.
type session struct {
	ctx     context.Context
	driver  *Driver
	bus     *driver.EventBus
	resume  driver.SessionID
	scripts []Script

	mu        sync.Mutex
	sessionID driver.SessionID
	scriptIdx int
	closed    bool
	active    *turn
	blocked   *driver.BlockedReason

	state atomic.Pointer[driver.ProcessState]
}

type turn struct {
	runID       driver.RunID
	script      Script
	stop        chan struct{}
	interrupted bool
}

func (s *session) SessionID() driver.SessionID {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.sessionID
}

func (s *session) ProcessState() driver.ProcessState {
	if p := s.state.Load(); p != nil {
		return *p
	}
	return driver.ProcessState{Phase: driver.PhaseIdle}
}

func (s *session) Raw() (driver.RawSessionData, error) {
	return driver.RawSessionData{}, driver.NewUnsupportedError("fake: raw session data are not supported")
}

func (s *session) TokenStatistics() (driver.TokenUsage, error) {
	return driver.TokenUsage{}, driver.NewUnsupportedError("fake: session token statistics are not supported")
}

func (s *session) setState(st driver.ProcessState) {
	s.state.Store(&st)
	s.bus.Emit(driver.LifecycleEvent(st))
}

// Start brings the session online: resume reuses the pre-assigned id, otherwise
// a fresh one is minted.
func (s *session) Start() error {
	if err := s.ctx.Err(); err != nil {
		return err
	}
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return driver.NewUnsupportedError("fake: session is closed")
	}
	if s.resume != "" {
		s.sessionID = s.resume
	} else {
		s.sessionID = s.driver.nextSessionID()
	}
	sid := s.sessionID
	s.mu.Unlock()

	s.setState(driver.ProcessState{Phase: driver.PhaseStarting})
	s.bus.Emit(driver.SessionAttachedEvent(sid))
	s.setState(driver.ProcessState{Phase: driver.PhaseActive, SessionID: sid})

	return nil
}

// Send replays the next scripted response (or a trivial echo when unscripted).
func (s *session) Send(input driver.UserInput) (driver.RunID, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return "", driver.NewUnsupportedError("fake: session is closed")
	}
	if err := s.ctx.Err(); err != nil {
		return "", err
	}
	sid := s.sessionID
	if sid == "" {
		return "", driver.NewProtocolError("fake: Send before Start")
	}
	if s.active != nil || s.blocked != nil {
		return "", driver.NewInvalidStateError("fake: session is not idle")
	}
	message, ok := input.(driver.UserMessage)
	if !ok {
		return "", driver.NewInvalidStateError("fake: a new turn requires UserMessage")
	}
	script, scripted := s.nextScriptLocked()
	if !scripted {
		script = Script{
			Items:  []driver.AgentEventItem{{Kind: driver.ItemText, Text: "echo: " + message.Text}},
			Result: driver.RunResult{FinishReason: driver.FinishNatural},
		}
	}
	return s.startScriptLocked(script), nil
}

func (s *session) nextScriptLocked() (Script, bool) {
	if s.scriptIdx >= len(s.scripts) {
		return Script{}, false
	}
	script := s.scripts[s.scriptIdx]
	s.scriptIdx++
	return script, true
}

func (s *session) startScriptLocked(script Script) driver.RunID {
	runID := driver.NewRunID()
	t := &turn{runID: runID, script: script, stop: make(chan struct{})}
	s.active = t
	s.setState(driver.ProcessState{Phase: driver.PhasePromptInFlight, SessionID: s.sessionID, RunID: runID})
	go s.execute(t)
	return runID
}

func (s *session) execute(t *turn) {
	if t.script.Wait != nil {
		select {
		case <-t.script.Wait:
		case <-t.stop:
			return
		}
	}

	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed || s.active != t || t.interrupted {
		return
	}
	for _, item := range t.script.Items {
		s.bus.Emit(driver.OutputEvent(s.sessionID, t.runID, item))
	}
	if t.script.Blocked != nil {
		reason := *t.script.Blocked
		s.active = nil
		s.blocked = &reason
		s.setState(driver.ProcessState{Phase: driver.PhaseBlocked, SessionID: s.sessionID, RunID: t.runID})
		s.bus.Emit(driver.BlockedEvent(s.sessionID, t.runID, reason))
		return
	}
	s.bus.Emit(driver.OutputEvent(s.sessionID, t.runID, driver.AgentEventItem{Kind: driver.ItemTurnEnd}))
	if t.script.FailWith != nil {
		s.bus.Emit(driver.FailedEvent(s.sessionID, t.runID, t.script.FailWith))
	} else {
		result := t.script.Result
		if result.FinishReason == "" {
			result.FinishReason = driver.FinishNatural
		}
		s.bus.Emit(driver.CompletedEvent(s.sessionID, t.runID, result))
	}
	s.active = nil
	s.setState(driver.ProcessState{Phase: driver.PhaseActive, SessionID: s.sessionID})
}

func (s *session) Respond(input driver.UserInput) (driver.RunID, error) {
	if err := s.ctx.Err(); err != nil {
		return "", err
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.closed {
		return "", driver.NewUnsupportedError("fake: session is closed")
	}
	if s.blocked == nil {
		return "", driver.NewInvalidStateError("fake: no blocked turn")
	}
	responseText := ""
	if len(s.blocked.Options) > 0 {
		selection, ok := input.(driver.OptionSelection)
		if !ok {
			return "", driver.NewInvalidStateError("fake: blocked options require OptionSelection")
		}
		valid := false
		for _, option := range s.blocked.Options {
			if option.Value == selection.Value {
				valid = true
				break
			}
		}
		if !valid {
			return "", driver.NewInvalidStateError("fake: invalid blocked response")
		}
		responseText = selection.Value
	} else {
		message, ok := input.(driver.UserMessage)
		if !ok {
			return "", driver.NewInvalidStateError("fake: blocked user input requires UserMessage")
		}
		responseText = message.Text
	}
	s.blocked = nil
	script, scripted := s.nextScriptLocked()
	if !scripted {
		script = Script{Items: []driver.AgentEventItem{{Kind: driver.ItemText, Text: "echo: " + responseText}}}
	}
	return s.startScriptLocked(script), nil
}

// Interrupt is an idempotent no-op unless a run is currently in flight.
func (s *session) Interrupt() error {
	ctx := s.ctx
	s.mu.Lock()
	if s.blocked != nil {
		s.blocked = nil
		s.setState(driver.ProcessState{Phase: driver.PhaseActive, SessionID: s.sessionID})
		s.mu.Unlock()
		return nil
	}
	t := s.active
	if t == nil {
		s.mu.Unlock()
		return nil
	}
	wait := t.script.InterruptWait
	s.mu.Unlock()
	if wait != nil {
		select {
		case <-wait:
		case <-ctx.Done():
			return ctx.Err()
		}
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.active != t {
		return nil
	}
	t.interrupted = true
	close(t.stop)
	s.active = nil
	s.bus.Emit(driver.OutputEvent(s.sessionID, t.runID, driver.AgentEventItem{Kind: driver.ItemTurnEnd}))
	s.bus.Emit(driver.CompletedEvent(s.sessionID, t.runID, driver.RunResult{FinishReason: driver.FinishCancelled}))
	s.setState(driver.ProcessState{Phase: driver.PhaseActive, SessionID: s.sessionID})
	return nil
}

// Close moves to PhaseClosed and closes the event bus. Idempotent.
func (s *session) Close() error {
	s.mu.Lock()
	if s.closed {
		s.mu.Unlock()
		return nil
	}
	s.closed = true
	s.mu.Unlock()
	s.setState(driver.ProcessState{Phase: driver.PhaseClosed, SessionID: s.SessionID()})
	s.bus.Close()
	return nil
}

// Compile-time interface checks.
var (
	_ driver.Driver  = (*Driver)(nil)
	_ driver.Session = (*session)(nil)
)

File: python/src/unio/_drivers/fake.py
from __future__ import annotations

import asyncio
import uuid
from collections import deque
from dataclasses import dataclass, field

from .._driver import (
    Driver,
    DriverEvent,
    DriverEventType,
    DriverSession,
    FinishReason,
    OutputItem,
    ProcessPhase,
    RunResult,
    StoredSession,
    new_run_id,
)
from ..errors import AgentError, invalid_state, unsupported
from ..models import (
    BlockedReason,
    OptionSelection,
    RawSessionData,
    TokenStatistics,
    UserInput,
    UserMessage,
)


@dataclass(slots=True)
class Script:
    items: list[OutputItem] = field(default_factory=lambda: [])
    result: RunResult = field(default_factory=RunResult)
    blocked: BlockedReason | None = None
    error: AgentError | None = None
    wait: asyncio.Event | None = None


class FakeDriver(Driver):
    def __init__(self, *scripts: Script) -> None:
        self._scripts: deque[Script] = deque(scripts)
        self._sessions: list[FakeSession] = []

    def probe(self) -> None:
        return

    async def list_sessions(self, cwd: str | None) -> list[StoredSession]:
        return [
            StoredSession(item.session_id, cwd=item.cwd)
            for item in self._sessions
            if item.session_id and (cwd is None or item.cwd == cwd)
        ]

    async def open_session(self, resume_id: str = "", cwd: str = "") -> DriverSession:
        session = FakeSession(self, resume_id, cwd)
        self._sessions.append(session)
        return session

    async def close(self) -> None:
        await asyncio.gather(*(item.close() for item in self._sessions))

    def next_script(self, prompt: str) -> Script:
        if self._scripts:
            return self._scripts.popleft()
        from .._driver import OutputKind

        return Script(items=[OutputItem(OutputKind.TEXT, text=f"echo: {prompt}")])


class FakeSession(DriverSession):
    def __init__(self, driver: FakeDriver, resume_id: str, cwd: str) -> None:
        super().__init__()
        self._driver = driver
        self._id = resume_id
        self.cwd = cwd
        self._phase = ProcessPhase.IDLE
        self._active_run = ""
        self._blocked: BlockedReason | None = None
        self._task: asyncio.Task[None] | None = None

    @property
    def session_id(self) -> str:
        return self._id

    @property
    def phase(self) -> ProcessPhase:
        return self._phase

    async def start(self) -> None:
        if self._phase is ProcessPhase.CLOSED:
            raise invalid_state("fake session is closed")
        if not self._id:
            self._id = f"fake-{uuid.uuid4()}"
        self._phase = ProcessPhase.ACTIVE
        self.events.emit(DriverEvent(DriverEventType.SESSION_ATTACHED, session_id=self._id))

    async def send(self, value: UserInput) -> str:
        if self._phase is not ProcessPhase.ACTIVE:
            raise invalid_state("fake session is not active")
        if not isinstance(value, UserMessage):
            raise invalid_state("fake: a new turn requires UserMessage")
        return self._start_script(self._driver.next_script(value.text))

    async def respond(self, value: UserInput) -> str:
        if self._phase is not ProcessPhase.BLOCKED or self._blocked is None:
            raise invalid_state("fake session is not blocked")
        if self._blocked.options:
            if not isinstance(value, OptionSelection):
                raise invalid_state("fake: blocked options require OptionSelection")
            if value.value not in {item.value for item in self._blocked.options}:
                raise invalid_state("invalid blocked response")
            response = value.value
        else:
            if not isinstance(value, UserMessage):
                raise invalid_state("fake: blocked user input requires UserMessage")
            response = value.text
        self._blocked = None
        return self._start_script(self._driver.next_script(response))

    async def interrupt(self) -> None:
        if self._phase is ProcessPhase.BLOCKED:
            self._blocked = None
            self._phase = ProcessPhase.ACTIVE
            return
        if self._phase is not ProcessPhase.PROMPT_IN_FLIGHT:
            return
        if self._task is not None:
            self._task.cancel()
            try:
                await self._task
            except asyncio.CancelledError:
                pass
        run_id = self._active_run
        self._active_run = ""
        self._phase = ProcessPhase.ACTIVE
        self.events.emit(
            DriverEvent(
                DriverEventType.COMPLETED,
                session_id=self._id,
                run_id=run_id,
                result=RunResult(FinishReason.CANCELLED),
            )
        )

    async def raw(self) -> RawSessionData:
        raise unsupported("fake session data is unsupported")

    async def token_statistics(self) -> TokenStatistics:
        raise unsupported("fake session statistics are unsupported")

    async def close(self) -> None:
        if self._phase is ProcessPhase.CLOSED:
            return
        if self._task is not None:
            self._task.cancel()
            await asyncio.gather(self._task, return_exceptions=True)
        self._phase = ProcessPhase.CLOSED
        self.events.close()

    def _start_script(self, script: Script) -> str:
        run_id = new_run_id()
        self._active_run = run_id
        self._phase = ProcessPhase.PROMPT_IN_FLIGHT
        self._task = asyncio.create_task(self._execute(run_id, script))
        return run_id

    async def _execute(self, run_id: str, script: Script) -> None:
        if script.wait is not None:
            await script.wait.wait()
        for item in script.items:
            self.events.emit(
                DriverEvent(
                    DriverEventType.OUTPUT,
                    session_id=self._id,
                    run_id=run_id,
                    item=item,
                )
            )
        self._active_run = ""
        if script.blocked is not None:
            self._blocked = script.blocked
            self._phase = ProcessPhase.BLOCKED
            self.events.emit(
                DriverEvent(
                    DriverEventType.BLOCKED,
                    session_id=self._id,
                    run_id=run_id,
                    blocked=script.blocked,
                )
            )
            return
        self._phase = ProcessPhase.ACTIVE
        if script.error is not None:
            self.events.emit(
                DriverEvent(
                    DriverEventType.FAILED,
                    session_id=self._id,
                    run_id=run_id,
                    error=script.error,
                )
            )
            return
        self.events.emit(
            DriverEvent(
                DriverEventType.COMPLETED,
                session_id=self._id,
                run_id=run_id,
                result=script.result,
            )
        )
